### ISSUE TITLE
Document interaction with http4s Router and relative paths

### ISSUE THREAD

**Issue by keynmol** [disneystreaming/smithy4s#152]
Let's say you define your smithy ops as mounted at root: 

```scala
@http(method: "POST", uri: "/{name}", code: 200)
```

If you then build the routes using `val routes = SimpleRestJsonBuilder.routes(...).resource`, you can't just shift the API to a custom prefix:

```scala
val shifted = Router("/api" -> routes)
```

The request to `/api/bla` will fail with 404.

This approach will work with manually defined routes, e.g. 

```scala 
    val other = HttpRoutes.of[IO] {
      case r if r.pathInfo.segments.headOption.exists(_.decoded() == "hello") =>
        import org.http4s.dsl.io.*
        Ok(r.toString)
    }

val shifted = Router("/api" -> (routes <+> other))
```

The request to `/api/hello/bla` will succeed.

I don't know exactly whether the recommendation should be writing absolute paths in the smithy files, or whether this should be considered undesired behaviour. Note that it will obviously affect the generated Swagger docs as well.

**Comment by Baccata** [comment]
> I don't know exactly whether the recommendation should be writing absolute paths in the smithy files 

That is a protocol-specific concern, and for the "simpleRestJson" protocol, the implicit rule is that the API spec should contain absolute paths. The rationale is as follows : 

1. It reinforces the source of truth aspect of the smithy specs 
2. It facilitates interop with other tooling (such as swagger/openapi) 

Facilitating shifting would break that paradigm. That being said, the fact that it doesn't work is somewhat accidental, and I'd like to understand why. But I'm not sure I want to fix it, since the bug is, in essence, a happy accident 😅. 

You are correct in the fact that it ought to be documented

**Comment by kubukoz** [comment]
https://smithy.io/2.0/spec/http-bindings.html?highlight=http#uri

> The resolved absolute URI of an operation is formed by combining the URI of the operation with the endpoint of the service. (that is, the host and any base URL of where the service is deployed). For example, given a service endpoint of https://example.com/v1 and an operation pattern of /myresource, the resolved absolute URI of the operation is https://example.com/v1/myresource.

It's fine that `@simpleRestJson` doesn't support this, but I think `SmithyHttp4sRouter` could. For other protocols which might want to reuse that, it'd save some trouble and still be in line with the Smithy spec (_I think_)

**Comment by tobiajo** [comment]
I agree this is very unexpected behavior. I'm considering to work around the issue like this:

```
httpRoutes = Router(
  "/prefix/not/in/smithy/uri" -> serverMiddleware(stripUriPrefix(smithyRoutes)),
  "/health" -> healthRoutes,
)

def stripUriPrefix[F[_]]: HttpRoutes[F] => HttpRoutes[F] = routes =>
  Kleisli { req =>
    routes.run(req.withUri(req.uri.withPath(req.pathInfo)))
  }
```

As I want to isolate routes and not leak the prefix, I see no alternative.
